connectors/s3-connector/src/main/scala/zio/connect/s3/S3Connector.scala

```diff
@@ -0,0 +1,74 @@
+package zio.connect.s3
+
+import zio.connect.s3.S3Connector.{CopyObject, MoveObject, S3Exception}
+import zio.stream.{ZSink, ZStream}
+import zio.{Trace, ZIO}
+
+trait S3Connector {
+
+  def copyObject(implicit trace: Trace): ZSink[Any, S3Exception, CopyObject, CopyObject, Unit]
+
+  def createBucket(implicit trace: Trace): ZSink[Any, S3Exception, String, String, Unit]
+
+  def deleteEmptyBucket(implicit trace: Trace): ZSink[Any, S3Exception, String, String, Unit]
+
+  def deleteObjects(bucketName: => String)(implicit
+    trace: Trace
+  ): ZSink[Any, S3Exception, String, String, Unit]
+
+  final def existsBucket(implicit trace: Trace): ZSink[Any, S3Exception, String, String, Boolean] =
+    ZSink
+      .take[String](1)
+      .map(_.headOption)
+      .mapZIO {
+        case Some(name) =>
+          for {
+            listResponse <- listBuckets.runCollect
+            bucketExists  = listResponse.contains(name)
+          } yield bucketExists
+        case None => ZIO.succeed(false)
+      }
+
+  final def existsObject(bucket: => String)(implicit trace: Trace): ZSink[Any, S3Exception, String, String, Boolean] =
+    ZSink
+      .take[String](1)
+      .map(_.headOption)
+      .mapZIO {
+        case Some(key) =>
+          for {
+            listResponse <- listObjects(bucket).runCollect
+            objectExists  = listResponse.contains(key)
+          } yield objectExists
+        case None => ZIO.succeed(false)
+      }
+
+  def getObject(bucketName: => String, key: String)(implicit trace: Trace): ZStream[Any, S3Exception, Byte]
+
+  def listBuckets(implicit trace: Trace): ZStream[Any, S3Exception, String]
+
+  def listObjects(bucketName: => String)(implicit trace: Trace): ZStream[Any, S3Exception, String]
+
+  def moveObject(implicit trace: Trace): ZSink[Any, S3Exception, MoveObject, MoveObject, Unit]
+
+  def putObject(bucketName: => String, key: String)(implicit trace: Trace): ZSink[Any, S3Exception, Byte, Nothing, Unit]
+
+}
+
+object S3Connector {
+
+  case class S3Exception(reason: Throwable)
+
+  final case class CopyObject(
+    sourceBucketName: String,
```

Comment:
zio-aws defines types for these primitive string values (backed by zio-prelude newtypes) such as `Bucket`, `S3Key` etc (not sure about the exact names). I think it would be nice to use those instead of pure strings for everything.

Reply:
I would not use the zio-aws api here avoid dependencies to that api. 
You are right about replacing the strings with newtypes, I will create some and update usages.

Reply:
I'm not convinced by this argument. I think it will cause tons of extra work to duplicate (parts of) the data model of each supported connector, and extra confusion in case somebody uses both zio-connect and the underlying library for some reason.

Reply:
I think it's worth it to keep them separate.
But we could help that mix case by providing an object with (implicit) conversions from connector.S3 types to zio-aws types. It could be imported with smth like`import zio.connect.s3.interop.zioaws._`

Reply:
Merging it like this for the hackathon we can revisit this later.